### ISSUE TITLE
Selecting GPU causes crash in VirtualBox

### ISSUE THREAD

**Issue by linuxlite** [hakandundar34coding/system-monitoring-center#24]
system-monitoring-center Version: 1.2.1

VirtualBox: 6.1.26_Ubuntu r145957

Output:

 
```
 jerry ~  system-monitoring-center

(system-monitoring-center:3757): Gdk-ERROR **: 16:35:09.645: The program 'system-monitoring-center' received an X Window System error.
This probably reflects a bug in the program.
The error was 'GLXBadFBConfig'.
  (Details: serial 4398 error_code 167 request_code 153 (GLX) minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap
```
![Screenshot_2022-02-05_16-41-03](https://user-images.githubusercontent.com/5382578/152627521-285244ec-9d5f-485d-824a-4c0e521f8a0b.png)

**Comment by hakandundar34coding** [comment]
I have tested the application on several distributions on VirtualBox but there was no error or a warning message during these tests. The test system has X11 windowing system and XFCE/GNOME DEs.

Can you share output of these commands and share screenshot of the GPU tab?

```glxinfo -B```
```DRI_PRIME=0 glxinfo -B```
```env DRI_PRIME=0 glxinfo -B```
```env DRI_PRIME=1 glxinfo -B```

**Comment by linuxlite** [comment]
```
 wwww  ~  glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0x15ad)
    Device: SVGA3D; build: RELEASE;  LLVM; (0x405)
    Version: 21.0.3
    Accelerated: no
    Video memory: 1MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: SVGA3D; build: RELEASE;  LLVM;
OpenGL version string: 2.1 Mesa 21.0.3
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 21.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

 wwww  ~  DRI_PRIME=0 glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0x15ad)
    Device: SVGA3D; build: RELEASE;  LLVM; (0x405)
    Version: 21.0.3
    Accelerated: no
    Video memory: 1MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: SVGA3D; build: RELEASE;  LLVM;
OpenGL version string: 2.1 Mesa 21.0.3
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 21.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

 wwww  ~  env DRI_PRIME=0 glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0x15ad)
    Device: SVGA3D; build: RELEASE;  LLVM; (0x405)
    Version: 21.0.3
    Accelerated: no
    Video memory: 1MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: SVGA3D; build: RELEASE;  LLVM;
OpenGL version string: 2.1 Mesa 21.0.3
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 21.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

 wwww  ~  env DRI_PRIME=1 glxinfo -B
name of display: :0.0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: VMware, Inc. (0x15ad)
    Device: SVGA3D; build: RELEASE;  LLVM; (0x405)
    Version: 21.0.3
    Accelerated: no
    Video memory: 1MB
    Unified memory: no
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: VMware, Inc.
OpenGL renderer string: SVGA3D; build: RELEASE;  LLVM;
OpenGL version string: 2.1 Mesa 21.0.3
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 21.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

 wwww  ~  

```

_....and share screenshot of the GPU tab?_

The title of the issue I thought was a dead giveaway.

**Comment by hakandundar34coding** [comment]
I am aware of the title. There may be errors in several steps. I did not know if the error is printed in the beginning of the tab (during loading the GUI) or 
before getting GPU driver information. I wanted to learn something about this by looking at the screenshot. GUI shows some of the information even if there are errors.

There is no problem on the output of ```glxinfo``` commands.

Can you download the file ```gpu_tab_debug.zip```, extract it, open a terminal in the same folder, run this command and write the output?
```ptyhon3 ./gpu_tab_debug.py```

[gpu_tab_debug.zip](https://github.com/hakandundar34coding/system-monitoring-center/files/8007804/gpu_tab_debug.zip)

**Comment by linuxlite** [comment]
Typo in your python command above.

```
python3 ./gpu_tab_debug.py 
<GdkX11.X11Screen object at 0x7f75bbf11540 (GdkX11Screen at 0x13cc050)>
./gpu_tab_debug.py:15: DeprecationWarning: Gdk.Screen.get_width is deprecated
  current_resolution = str(current_screen.get_width()) + "x" + str(current_screen.get_height())
./gpu_tab_debug.py:15: DeprecationWarning: Gdk.Screen.get_height is deprecated
  current_resolution = str(current_screen.get_width()) + "x" + str(current_screen.get_height())
1920x1200

```

```
system-monitoring-center

(system-monitoring-center:2693): Gdk-ERROR **: 00:18:41.801: The program 'system-monitoring-center' received an X Window System error.
This probably reflects a bug in the program.
The error was 'GLXBadFBConfig'.
  (Details: serial 59516 error_code 167 request_code 153 (GLX) minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/breakpoint trap
```

**Comment by linuxlite** [comment]
FYI - I'm look for application to replace both htop and task manager in Series 6x
But it has to be rock solid, including in VM's before I'll consider adding it. Cheers :)

**Comment by hakandundar34coding** [comment]
There is no error in the output of the script file. There are some warnings.

I will reinstall Ubuntu on VirtualBox and try to reproduce the problem. I will install 20.04.3 LTS. You can write here if you want to specify details about the system which should to be installed. Some details can be found from the screenshot of the System tab.

Additional information about GPU tab:
Information about GPU on systems with:
Single GPU can be get.
Multiple GPUs with one open sourced and one closed sourced driver can be get.
But this information may not be get on other GPU/driver configurations.

Multiple GPUs in this comment means that integrated GPU + discrete GPU.

I did not write any code to get information to get information by using closed sourced driver programs. There are some GPU/software configurations that ```DRI_PRIME=1 glxinfo -B``` does not work without problems on them.

But device and model information can be get for all GPUs. Situations above are valid for information on the right column on the GPU tab.

**Comment by hakandundar34coding** [comment]
There is no problem in the code of System Monitoring Center. The problem is in one of the libraries. Probably it is due to this bug in GTK and it looks like it is fixed in Gtk 3.24.23 but Ubuntu 20.04 uses Gtk 3.24.20.
https://bugs.launchpad.net/ubuntu/+source/gtk+3.0/+bug/1894737

I have tested the application on several systems.

GPU tab is working:
```
Ubuntu 20.04 - X11 windowing system - VirtualBox - Enable 3D Acceleration option is disabled (VirtualBox setting)
Pardus 21 (Debian-based distribution, kernel 5.10) - X11 windowing system - VirtualBox - Enable 3D Acceleration option is enabled (VirtualBox setting)
Pardus 21 (Debian-based distribution, kernel 5.10) - X11 windowing system - VirtualBox - Enable 3D Acceleration option is disabled (VirtualBox setting)
Ubuntu 20.04 - Wayland windowing system - Gnome Boxes (virtual machine)
```

GPU tab is not working:
```
Ubuntu 20.04 - Wayland windowing system - VirtualBox - Enable 3D Acceleration option is enabled (VirtualBox setting)
Ubuntu 20.04 - X11 windowing system - VirtualBox - Enable 3D Acceleration option is enabled (VirtualBox setting)
```


GPU tab is not shown on Ubuntu if Enable 3D Acceleration option is enabled on VirtualBox.
It crashes when one of the GUI objects of the GPU tab is loaded. This GUI object is `Gtk.GLArea`. This widget is used for counting FPS by showing graphics and it uses OpenGL for showing graphics.

Gtk GLArea test does not work on the same system (Ubuntu-VirtualBox-Enable 3D Acceleration) but works on Pardus 21 (Gtk 3.24.24).
The command to run this test:
```gtk3-demo --run=glarea```
This package is required for this test:
```sudo apt-get install gtk-3-examples```

I will close this issue because it is not a bug of System Monitoring Center. You can write comments here or open a new issue if there are another bugs.
